src/desktop/DesktopSqlCipher.ts

```diff
@@ -0,0 +1,142 @@
+		return assertNotNull(this._db)
+	}
+
+	/**
+	 * @param nativeBindingPath the path to the sqlite native module
+	 */
+	constructor(
+		private readonly nativeBindingPath: string,
+		private readonly dbPath: string,
+		private readonly integrityCheck: boolean,
+	) {
+	}
+
+
+	async openDb(userId: string, dbKey: Uint8Array): Promise<void> {
+		this._db = new sqlite(this.dbPath, {
+			// Remove ts-ignore once proper definition of Options exists, see https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/59049#
+			// @ts-ignore missing type
+			nativeBinding: this.nativeBindingPath,
+			// verbose: (message, args) => {
+			// 	console.log("DB", message, args)
+			// }
+		})
+		this.initSqlcipher({databaseKey: dbKey, enableMemorySecurity: true, integrityCheck: this.integrityCheck})
+		// FIXME: move me away
+		this.function("firstIdBigger", (l, r) => {
+			return boolToSqlite(firstBiggerThanSecond(l, r))
+		})
+		this.function("firstIdBiggerOrEq", (l, r) => {
+			return boolToSqlite(l == r || firstBiggerThanSecond(l, r))
+		})
+	}
+
+	async closeDb(): Promise<void> {
+		this.db.close()
+	}
+
+	deleteDb(userId: string): Promise<void> {
+		throw new Error("Not implemented")
+	}
+
+	/**
+	 * Initialise sqlcipher with a database key, configuration:
+	 * - Sqlcipher always uses aes-256 for encryption.
+	 * - Sqlcipher always creates per page hmac for integrity with sha512.
+	 * - Sqlcipher generates a database salt value randomly and stores in the first 16 bytes of the database.
+	 * - We pass the database key directly to sqlcipher rather than using a password and therefore do not configure key derivation.
+	 * - we assume that adding entropy to entropy pool of the crypto provide (cipher_add_random) "is not necessary [...], since [openssl] does (re-)seed itself automatically using trusted system entropy sources", https://www.openssl.org/docs/man1.1.1/man3/RAND_add.html
+	 * @param databaseKey
+	 * @param enableMemorySecurity if true the the memory security option (that was default until 4.5, https://www.zetetic.net/blog/2021/10/28/sqlcipher-4.5.0-release/) to wipe memory allocated by SQLite internally, including the page cache is enabled.
+	 * @param integrityCheck: if true the hmac stored with each page of the database is verified to detect modification.
+	 * @throws if an error is detected during the integrity check
+	 */
+	private initSqlcipher(
+		{databaseKey, enableMemorySecurity, integrityCheck}: {
+			databaseKey: Uint8Array,
+			enableMemorySecurity: boolean,
+			// integrity check breaks tests
+			integrityCheck: boolean
+		}
+	) {
+		if (enableMemorySecurity) {
+			this.db.pragma("cipher_memory_security = ON")
+		}
+		const key = `x'${uint8ArrayToBase64(databaseKey)};`
+		this.db.pragma(`KEY = "${key}"`)
+
+		if (integrityCheck) {
+			this.checkIntegrity()
+		}
+	}
+
+	/**
+	 * Define a function that can be used in queries
+	 */
+	function(functionName: string, functionCallback: (...params: any[]) => any) {
```

Comment:
We should just inline this or even remove it since we are doing the ID comparison inline because of android